room/compiler/src/main/kotlin/androidx/room/processor/Context.kt

```diff
@@ -37,7 +38,9 @@ class Context private constructor(
     private val typeConverters: CustomConverterProcessor.ProcessResult,
     private val inheritedAdapterStore: TypeAdapterStore?,
     val cache: Cache,
-    private val canRewriteQueriesToDropUnusedColumns: Boolean
+    private val canRewriteQueriesToDropUnusedColumns: Boolean,
+    // currently processed database element to type converter factory map
+    private val typeConverterFactoriesMap: MutableMap<XTypeElement, MutableSet<XType>>
```

Comment:
i'm not sure why we are keeping this in the context but i need to check more in detail (going into a meeting now :) )

Reply:
ok i guess this is here because type converters were initially kept in here because they are scoped to whereever the annotation is used.
If possible, it feels like it would be cleaner if the database processor collected each `used` converter from its DAO's and entities then keep that list in the `Database` value object (to verify later in the generated code that all used ones are provided).

Reply:
So since it's not possible to give a compile time validation for converter factories, I wanted to perform 2 checks at runtime (when `DatabaseBuilder.build()` is called). Those checks are:

1.  User added `@TypeConverterFactory(FooFactory.class)` on a type converter class but forgot to add the instance of `FooFactory` to database builder.

2. User provided an instance of `FooFactory` to database builder but there is no type converter class annotated with `@TypeConverterFactory(FooFactory.class)`.

In order to do that I needed an information about all `@TypeConverterFactory` annotations to be available in `RoomDatabase` file. I figured out that I can use `DatabaseWriter` to put this info there. `Context` already keeps track of type converters but it's a hierarchical data structure and at `DatabaseProcessor` level I have access only to the top level context. It's not possible to extract this information having access only to this context so I added a "flat" data structure that keeps track of all type converter factories.
Not sure what does it mean that a converter is being `used` or not but currently factories are added to a set inside a context. The logic of adding the factory to this set is in `CustomConverterProcessor` file. I'm not sure if this guarantees that a converter is being `used`. It was the only place I've found that allowed me to implement this feature. I'm open to suggestions though :)

Reply:
I think we want to move the set of factories to the value object we have to represent a processed database: `androidx.room.vo.Database`. I do still think we need the `context` to gather them since they can be scoped to entity fields, daos, even POJOs so its not easy to gather those from the `DatabaseProcessor`. But what we can do is keep it more local by creating the map of factories in `DatabaseProcessor` and clearing it from the context once we are done and have created a `androidx.room.vo.Database`. It also means we can remove the mutable property `databaseElement` that was added. `DaoWritter` would also stay the same, since the info is in the value object.

Reply:
I implemented @danysantiago suggestions. Let me know if it's fine now :)